modules/appgw/variables.tf

```diff
@@ -1,169 +1,632 @@
+  NOTICE: this is also enforced on the Public IP. The Public IP object brings in some limitations as it can only be non-zonal,
+  pinned to a single zone or zone-redundant (so available in all zones in a region).
+  Therefore make sure that if you specify more than one zone you specify all available in a region. You can use a subset,
+  but the Public IP will be created in all zones anyway. This fact will cause terraform to recreate the IP resource during
+  next `terraform apply` as there will be difference between the state and the actual configuration.
 
-  For details on zones currently available in a region of your choice refer to [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/availability-zones/az-region).
+  For details on zones currently available in a region of your choice refer to
+  [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/availability-zones/az-region).
 
   Example:
   ```
   zones = ["1","2","3"]
   ```
   EOF
-  default     = null
+  default     = ["1", "2", "3"]
   type        = list(string)
+  validation {
+    condition     = var.zones == null || length(setsubtract(var.zones, ["1", "2", "3"])) == 0
+    error_message = "The `var.zones` can either bea non empty list of Availability Zones or explicit `null`."
+  }
 }
 
-variable "name" {
-  description = "Name of the Application Gateway."
+variable "public_ip_name" {
+  description = "Name for the public IP address."
   type        = string
 }
 
 variable "domain_name_label" {
-  description = "Label for the Domain Name. Will be used to make up the FQDN. If a domain name label is specified, an A DNS record is created for the public IP in the Microsoft Azure DNS system."
+  description = <<-EOF
+  Label for the Domain Name.
+
+  Will be used to make up the FQDN. If a domain name label is specified, an A DNS record is created
+  for the public IP in the Microsoft Azure DNS system."
+  EOF
   default     = null
   type        = string
 }
 
+variable "enable_http2" {
+  description = "Enable HTTP2 on the Application Gateway."
+  default     = false
+  nullable    = false
+  type        = bool
+}
+
+variable "waf" {
+  description = <<-EOF
+  Object sets only the SKU and provide basic WAF configuration for Application Gateway.
+
+  This module does not support WAF rules configuration and advanced WAF settings.
+  Only below attributes are allowed:
+  - `enabled`          - (`bool`, required) Enables WAF Application Gateway
+  - `firewall_mode`    - (`string`, optional) The Web Application Firewall Mode
+  - `rule_set_type`    - (`string`, optional, defaults to `OWASP`) The Type of the Rule Set used for this Web Application Firewall
+  - `rule_set_version` - (`string`, optional) The Version of the Rule Set used for this Web Application Firewall
+  EOF
+  default = {
+    enabled = false
+  }
+  nullable = false
+  type = object({
+    enabled          = bool
+    firewall_mode    = optional(string)
```

Comment:
maybe set this to a bool value? something like `prevention_mode = true/false`? it's highly unlikely that we will have a 3rd option, and maybe this would make this less error-prone (type-o's, etc)

Reply:
We have validation, so there will be no errors in deployment:
```
  validation {
    condition     = contains(["Detection", "Prevention"], coalesce(var.waf.firewall_mode, "Detection"))
    error_message = "For `firewall_mode` possible values are Detection and Prevention"
  }
```

I prefer to have it as string, because this is the way how it's defined in https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/application_gateway.

Can we leave it as it is ?